refactor(options)!: use forward slash as path separator on Windows

famiu opened the pull request:
Problem: All modern Windows versions support using a forward slash as path separator on most applications, including the two most common shells, cmd.exe and PowerShell. Therefore, using backward slash as path separator is largely useless.

Solution: Use forward slash as path separator on every platform. Also remove `completeslash` and `shellslash` as they are no longer needed.

famiu committed:
refactor(options)!: use forward slash as path separator on Windows

Problem: All modern Windows versions support using a forward slash as path separator on most applications, including the two most common shells, cmd.exe and PowerShell. Therefore, using backward slash as path separator is largely useless.

Solution: Use forward slash as path separator on every platform. Also remove `completeslash` and `shellslash` as they are no longer needed.

zeertzjq commented on src/nvim/path.c:
```diff
@@ -1409,38 +1409,6 @@ static int expand_backtick(garray_T *gap, char *pat, int flags)
-/// backslash twice.
-/// When 'shellslash' set do it the other way around.
-/// When the path looks like a URL leave it unmodified.
-void slash_adjust(char *p)
```
Why is this removed? Won't that cause backslashes in paths return by libuv to remain backslashes?

clason replied:
That assumes that libuv returns backslashes, which afaik is not the case.

zeertzjq replied:
I just tested with both luv and FFI on Windows. It is the case.

zeertzjq reviewed (changes requested):
This should make `exists('+shellslash')` always return `false`, which isn't the case with `immutable = true`. There are two options:
- Use `enable_if = false` instead if `immutable = true`.
- Make `exists('+option')` return `false` for options with `immutable = true`.

famiu committed:
modify `slash_adjust()`

famiu committed:
refactor(options): replace `enable_if = false` with `hidden = true`

Problem: Currently, `enable_if = false` is used to mark an option as hidden, instead of the pre-existing `hidden` flag. Options hidden with `enable_if = false` also have their variable pointer set to `NULL`, which means their value can't be fetched using Vim script or the API which is also unintended behavior.

Solution: Use `hidden = true` to mark an option as hidden, make hidden option variable pointers always point to the default value, and also allow fetching the value of a hidden option using Vim script and the API.

famiu committed:
remove references to `p_csl` and `p_ssl`

famiu committed:
remove more references to shellslash

justinmk commented on runtime/doc/builtin.txt:
```diff
@@ -6957,10 +6957,8 @@ sha256({string})                                                      *sha256()*
-		Otherwise encloses {string} in single-quotes and replaces all
-		"'" with "'\''".
+		On Windows, encloses {string} in single-quotes and replaces
+		all "'" with "'\''".
```
This is not correct. If the shell is cmd.exe, double quotes should be used (and "doubles all double-quotes"). shellescape() inherently depends on 'shell'.

zeertzjq commented:
vim-go uses 'noshellslash': https://github.com/fatih/vim-go/blob/ac1fb9611de435d80a6badcf8482cb1428449b10/plugin/go.vim#L105

famiu commented:
Superseded by #28400
